agents/tools/upstream_tools.py

```diff
@@ -0,0 +1,702 @@
+    is_pr: bool = Field(default=False, description="True if this is a pull request or merge request URL")
+
+
+class ExtractUpstreamRepositoryOutput(JSONToolOutput[UpstreamRepository]):
+    pass
+
+
+class ExtractUpstreamRepositoryTool(Tool[ExtractUpstreamRepositoryInput, ToolRunOptions, ExtractUpstreamRepositoryOutput]):
+    name = "extract_upstream_repository"
+    description = """
+    Extract upstream repository URL and commit hash from a commit or pull request URL.
+
+    Supports common formats:
+    - GitHub/GitLab commit: https://domain.com/owner/repo/commit/hash or /-/commit/hash
+    - GitHub/GitLab PR: https://domain.com/owner/repo/pull/123 or /merge_requests/123
+    - Query param formats: ?id=hash or ?h=hash (for cgit/gitweb)
+
+    For pull requests, fetches the head commit SHA from the PR.
+    Returns the git clone URL and commit hash needed for cherry-picking.
+    """
+    input_schema = ExtractUpstreamRepositoryInput
+
+    def _create_emitter(self) -> Emitter:
+        return Emitter.root().child(
+            namespace=["tool", "upstream", self.name],
+            creator=self,
+        )
+
+    async def _run(
+        self, tool_input: ExtractUpstreamRepositoryInput, options: ToolRunOptions | None, context: RunContext
+    ) -> ExtractUpstreamRepositoryOutput:
+        try:
+            parsed = urlparse(tool_input.upstream_fix_url)
+
+            # Check if this is a pull request URL
+            pr_match = re.search(r'/pull/(\d+)(?:\.patch)?', parsed.path)
+            mr_match = re.search(r'/merge_requests/(\d+)(?:\.patch)?', parsed.path)
+
+            if pr_match or mr_match:
+                # Handle GitHub Pull Request or GitLab Merge Request
+                pr_number = pr_match.group(1) if pr_match else mr_match.group(1)
+
+                # Extract owner/repo from path
+                if pr_match:
+                    repo_path_match = re.match(r'/([\w\-\.]+)/([\w\-\.]+)/pull/', parsed.path)
+                else:
+                    repo_path_match = re.match(r'/([\w\-\.]+)/([\w\-\.]+)/-/merge_requests/', parsed.path)
+
+                if not repo_path_match:
+                    raise ToolError(f"Could not extract repository path from PR/MR URL: {tool_input.upstream_fix_url}")
+
+                owner, repo = repo_path_match.group(1), repo_path_match.group(2)
+
+                # Fetch PR/MR information to get the head commit
+                if pr_match:
+                    # GitHub API
+                    api_url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{pr_number}"
+                else:
+                    # GitLab API
+                    api_url = f"https://{parsed.netloc}/api/v4/projects/{owner}%2F{repo}/merge_requests/{pr_number}"
+
+                try:
+                    req = Request(api_url)
+                    req.add_header('Accept', 'application/json')
+                    req.add_header('User-Agent', 'RHEL-Backport-Agent')
```

Comment:
for the API stuff, we could (probably later) use our https://github.com/packit/ogr library to simplify few things here